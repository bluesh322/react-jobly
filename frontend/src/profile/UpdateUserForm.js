import React, { useState, useContext } from "react";
import clsx from "clsx";
import { useHistory } from "react-router-dom";
import {
  Typography,
  Container,
  Box,
  FormControl,
  TextField,
  Button,
  makeStyles,
  Card,
  CardContent,
} from "@material-ui/core";
import { Alert } from "@material-ui/lab"
import UserContext from "../auth/UserContext";
import JoblyApi from "../api";

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    flexWrap: "wrap",
    justifyContent: "center",
  },
  margin: {
    margin: theme.spacing(1),
  },
  withoutLabel: {
    marginTop: theme.spacing(3),
  },
  textField: {
    width: "25ch",
  },
  btn: {
    width: 200,
    marin: theme.spacing(2),
  },
  text: {
    display: "flex",
  },
}));

const UpdateUserForm = () => {
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const classes = useStyles();
  const INITIAL_STATE = {
    username: currentUser.username,
    password: "",
    firstName: currentUser.firstName,
    lastName: currentUser.lastName,
    email: currentUser.email,
  };
  const history = useHistory();
  const [formData, setFormData] = useState(INITIAL_STATE);
  const [formErrors, setFormErrors] = useState([]);
  const [saveConfirmed, setSaveConfirmed] = useState(false);

  const handleChange = (e) => {
    e.preventDefault();
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const profileData = {
      password: formData.password,
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
    };
    const {username} = currentUser;
    const {password} = formData;
    let updatedUser;
    try {
      let token = await JoblyApi.login({username, password});
      if(!token) return;
      updatedUser = await JoblyApi.saveProfile(username, profileData);
    } catch (errors) {
      //debugger;
      setFormErrors(errors);
      return;
    }
    setCurrentUser(updatedUser);
    setFormData(f => ({ ...f, password: ""}));
    setFormErrors([]);
    setSaveConfirmed(true);
    history.push("/profile");
  };

  return (
    <Container fullwidth="md" fixed>
      <Typography className={classes.root}  variant="h4">Profile</Typography>
      <Card>
        <CardContent>
          <Box className={classes.root} my={2}>
            <Typography className={classes.text} variant="h3">
            </Typography>
            <form onSubmit={handleSubmit}>
              <FormControl>
                <TextField
                  disabled
                  className={clsx(
                    classes.margin,
                    classes.withoutLabel,
                    classes.textField
                  )}
                  variant="outlined"
                  label="username"
                  name="username"
                  value={formData.username}
                  onChange={handleChange}
                ></TextField>
                <TextField
                  className={clsx(
                    classes.margin,
                    classes.withoutLabel,
                    classes.textField
                  )}
                  variant="outlined"
                  label="firstName"
                  name="firstName"
                  value={formData.firstName}
                  onChange={handleChange}
                ></TextField>
                <TextField
                  className={clsx(
                    classes.margin,
                    classes.withoutLabel,
                    classes.textField
                  )}
                  variant="outlined"
                  label="lastName"
                  name="lastName"
                  value={formData.lastName}
                  onChange={handleChange}
                ></TextField>
                <TextField
                  className={clsx(
                    classes.margin,
                    classes.withoutLabel,
                    classes.textField
                  )}
                  variant="outlined"
                  label="email"
                  name="email"
                  value={formData.email}
                  onChange={handleChange}
                ></TextField>
                <TextField
                  className={clsx(
                    classes.margin,
                    classes.withoutLabel,
                    classes.textField
                  )}
                  variant="outlined"
                  label="Confirm password"
                  name="password"
                  value={formData.password}
                  onChange={handleChange}
                ></TextField>
              {formErrors.length
                  ? <Alert severity="error">{formErrors}</Alert>
                  : null}

              {saveConfirmed
                  ?
                  <Alert severity="success" >Updated successfully.</Alert>
                  : null}

                <Button
                  className={classes.btn}
                  variant="contained"
                  color="primary"
                  type="submit"
                >
                  Update Profile
                </Button>
              </FormControl>
            </form>
          </Box>
        </CardContent>
      </Card>
    </Container>
  );
};

export default UpdateUserForm;
