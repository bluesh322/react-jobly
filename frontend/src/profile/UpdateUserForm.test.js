import React from "react";
import { render } from "@testing-library/react";
import UpdateUserForm from "./UpdateUserForm";
import TestUserProvider from "../auth/testUser";

// TODO: woefully under-tested!

it("matches snapshot", function () {
  const { asFragment } = render(
    <TestUserProvider>
      <UpdateUserForm />
    </TestUserProvider>
  );
  expect(asFragment()).toMatchSnapshot();
});
