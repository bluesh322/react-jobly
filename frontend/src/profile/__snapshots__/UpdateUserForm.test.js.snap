// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-fixed MuiContainer-maxWidthLg"
    fullwidth="md"
  >
    <h4
      class="MuiTypography-root makeStyles-root-1 MuiTypography-h4"
    >
      Profile
    </h4>
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <div
          class="MuiBox-root MuiBox-root-7 makeStyles-root-1"
        >
          <h3
            class="MuiTypography-root makeStyles-text-6 MuiTypography-h3"
          />
          <form>
            <div
              class="MuiFormControl-root"
            >
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-withoutLabel-3 makeStyles-textField-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                  data-shrink="true"
                >
                  username
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                    disabled=""
                    name="username"
                    type="text"
                    value="testuser"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10 PrivateNotchedOutline-legendNotched-11"
                    >
                      <span>
                        username
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-withoutLabel-3 makeStyles-textField-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                  data-shrink="true"
                >
                  firstName
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    name="firstName"
                    type="text"
                    value="test"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10 PrivateNotchedOutline-legendNotched-11"
                    >
                      <span>
                        firstName
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-withoutLabel-3 makeStyles-textField-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                  data-shrink="true"
                >
                  lastName
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    name="lastName"
                    type="text"
                    value="test"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10 PrivateNotchedOutline-legendNotched-11"
                    >
                      <span>
                        lastName
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-withoutLabel-3 makeStyles-textField-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled"
                  data-shrink="true"
                >
                  email
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    name="email"
                    type="text"
                    value="t@t.com"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10 PrivateNotchedOutline-legendNotched-11"
                    >
                      <span>
                        email
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <div
                class="MuiFormControl-root MuiTextField-root makeStyles-margin-2 makeStyles-withoutLabel-3 makeStyles-textField-4"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
                  data-shrink="false"
                >
                  Confirm password
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input"
                    name="password"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="PrivateNotchedOutline-root-8 MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="PrivateNotchedOutline-legendLabelled-10"
                    >
                      <span>
                        Confirm password
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-btn-5 MuiButton-containedPrimary"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Update Profile
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
